temporalio/lib/temporalio/worker/activity_executor/thread_pool.rb

```diff
@@ -0,0 +1,254 @@
+        def shutdown
+          @mutex.synchronize do
+            # Stop all workers
+            @pool.each(&:stop)
+          end
+        end
+
+        # Kill each thread. This should not be called until all workers using this executor are complete. This does not
+        # need to be called at all on program exit (e.g. for the global default).
+        def kill
+          @mutex.synchronize do
+            # Kill all workers
+            @pool.each(&:kill)
+            @pool.clear
+            @ready.clear
+          end
+        end
+
+        # @!visibility private
+        def _remove_busy_worker(worker)
+          @mutex.synchronize { locked_remove_busy_worker(worker) }
+        end
+
+        # @!visibility private
+        def _ready_worker(worker, last_message)
+          @mutex.synchronize { locked_ready_worker(worker, last_message) }
+        end
+
+        # @!visibility private
+        def _worker_died(worker)
+          @mutex.synchronize { locked_worker_died(worker) }
+        end
+
+        # @!visibility private
+        def _worker_task_completed
+          @mutex.synchronize { @completed_task_count += 1 }
+        end
+
+        private
+
+        def locked_assign_worker(&block)
+          # keep growing if the pool is not at the minimum yet
+          worker, = @ready.pop || locked_add_busy_worker
+          if worker
+            worker << block
+            true
+          else
+            false
+          end
+        end
+
+        def locked_enqueue(&block)
+          @queue << block
+        end
+
+        def locked_add_busy_worker
+          return if @max_threads && @pool.size >= @max_threads
+
+          @workers_counter += 1
+          @pool << (worker = Worker.new(self, @workers_counter))
+          @largest_length = @pool.length if @pool.length > @largest_length
+          worker
+        end
+
+        def locked_prune_pool
+          now = ThreadPool._monotonic_time
+          stopped_workers = 0
+          while !@ready.empty? && (@pool.size - stopped_workers).positive?
+            worker, last_message = @ready.first
+            break unless now - last_message > @idle_timeout
+
+            stopped_workers += 1
+            @ready.shift
+            worker << :stop
+
+          end
+
+          @next_prune_time = ThreadPool._monotonic_time + @prune_interval
+        end
+
+        def locked_remove_busy_worker(worker)
+          @pool.delete(worker)
+        end
+
+        def locked_ready_worker(worker, last_message)
+          block = @queue.shift
+          if block
+            worker << block
+          else
+            @ready.push([worker, last_message])
+          end
+        end
+
+        def locked_worker_died(worker)
+          locked_remove_busy_worker(worker)
+          replacement_worker = locked_add_busy_worker
+          locked_ready_worker(replacement_worker, ThreadPool._monotonic_time) if replacement_worker
+        end
+
+        # @!visibility private
+        class Worker
```

Comment:
I found this a bit confusing with our overall `Worker` name. Maybe `PoolWorker` might be nice here? Not a big problem either way

Reply:
It's private to this class and in Ruby names of things are often ambiguous if viewed in only their unqualified form.

Reply:
Yeah, I understand that logically, but, it was still not immediately obvious to me